ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastDownloaderFactory.java

```diff
@@ -79,7 +80,9 @@
           "Fast sync was requested, but cannot be enabled because the local blockchain is not empty.");
       return Optional.empty();
     }
-
+    if (worldStateStorage instanceof BonsaiPersistedWorldState) {
+      worldStateStorage.clear();
```

Comment:
Indeed it is a workaround waiting to find better. There are two things we download the blocks headers and the recent state trie. The only part we reset is the recent state trie. The headers are not removed and it is a big part so we reduce the impact to the minimum. And the pivot block remains the same .

This is a fix to avoid that a person who restarts his node during a fastsync does not waste time synchronizing a corrupted node. But the typical use case is not to re-start a node during a fastsync 🧑‍🏭